src/Storages/MergeTree/MergeTreeIndexGin.cpp

```
@@ -758,44 +763,100 @@ MergeTreeIndexConditionPtr MergeTreeIndexGin::createIndexCondition(const Actions
     return std::make_shared<MergeTreeIndexConditionGin>(predicate, context, index.sample_block, gin_filter_params, token_extractor.get());
 }
 
-MergeTreeIndexPtr ginIndexCreator(const IndexDescription & index)
+namespace
+{
+[[nodiscard]] std::unordered_map<String, Field> convert_arguments_to_options_map(const FieldVector & arguments)
 {
-    size_t ngram_length = index.arguments.empty() ? 0 : index.arguments[0].safeGet<size_t>();
-    UInt64 max_rows = index.arguments.size() < 2 ? DEFAULT_MAX_ROWS_PER_POSTINGS_LIST : index.arguments[1].safeGet<UInt64>();
-    GinFilterParameters gin_filter_params(ngram_length, max_rows);
+    std::unordered_map<String, Field> options;
+    for (const Field & argument : arguments)
+    {
+        if (argument.getType() != Field::Types::Tuple)
+            throw Exception(ErrorCodes::INCORRECT_QUERY, "Arguments of GIN index must be key-value pair (identifier = literal)");
+        Tuple tuple = argument.template safeGet<Tuple>();
+        String key = tuple[0].safeGet<String>();
+        options[key] = tuple[1];
+    }
+    return options;
+}
 
-    if (ngram_length == 0)
+template <typename Type>
+[[nodiscard]] std::optional<Type> extract_option(const std::unordered_map<String, Field> & options, const String & option)
+{
+    if (auto it = options.find(option); it != options.end())
```

Let's throw an exception if the same key occurs > 1x.